### PR TITLE
Productionize Gemma4 E4B QAT Q4_0 on Metal: kernels, frame safety, sampling, speculative groundwork

### PR DESCRIPTION
**Pull request by timkaye11**
## Summary

Takes the Gemma4 E4B QAT Q4_0 Metal path from prototype to production: a
kernel-optimization pass across decode/prefill, a frame-safety fix for a
write-after-free class, temperature sampling rebuilt at greedy parity, and
device-resident speculative-decoding groundwork. Every change is gated on
exact-token golden checks; speed is tracked as same-run ratios vs llama.cpp
(fanless M4 throttles, so ratios are the honest metric).

| Metric | Result |
|---|---|
| Greedy decode | 24.5 tok/s — 0.91× llama.cpp same-run (26.9) |
| End-to-end generate (21-tok prompt + 128 tok) | 5.5 s — 0.90× llama.cpp |
| Sampled decode (temperature > 0) | 25.8 tok/s — 95% of greedy |
| Prefill | 567 ms @ 111-token prompt (~3.4 ms/tok marginal, llama pp parity) |
| MTP draft step | ~11 ms, fully device-resident |

## Kernels & decode pipeline

- q6_K LM-head mmv rewrite (ggml-style, ~99 GB/s) and q4_0 simdgroup GEMM for
  prefill (rows ≥ 9); wide-16 q4_0 small-batch kernel for rows 2–8.
- Prefill attention generalized to a multi-query grid; paged-attention
  value-phase made branchless; sliding-window layers now skip whole KV pages
  outside the window (bit-exact, win grows with context).
- Pipelined lag-by-one decode frames default-on (CPU encode fully hidden);
  f16 activation paths (PLE, split gate/up) default-on.
- Opt-in simdgroup flash-attention prefill kernel — correct but not yet faster
  than the tuned scalar kernel; off pending Instruments tuning.

## Sampling at greedy parity

- Sampled decode reuses the planned greedy frame and samples from the logits
  its fused tail already leaves resident (no second LM-head pass), via a new
  parallel Gumbel-max kernel with in-kernel softcap and penalties — replacing
  a single-thread 262k-element CDF walk.

## Speculative decoding (groundwork)

- Draft model fully device-resident (donated-KV slot attention, device LM-head
  argmax, framed forward). Identity gates exact (plain == force-k4 ==
  force-k1); still slower than plain decode (acceptance-bound) and off by
  default — roadmap documented.

## Validation

- `zig build test`: 1944 passed / 12 skipped / 0 failed.
- Compare gate exit 0 with exact 128-token golden IDs; MTP identity exact;
  sampled output diverse, near-zero temperature reproduces greedy.
- All new behavior has kill switches (`TERMITE_METAL_DISABLE_SWA_SCAN_CLAMP`,
  `..._Q4_0_SMALL_BATCH_WIDE`, `..._GATED_FFN_RESIDUAL_SINGLE_ROW`,
  `ANTFLY_GEMMA4_MTP_DISABLE_DRAFT_FRAME`); the experimental prefill kernel is
  opt-in (`TERMITE_METAL_ENABLE_PREFILL_SG_ATTENTION`).

### CODE REVIEW

File: zig/pkg/inference/src/architectures/gemma4_mtp.zig [hunk 1]
```diff
@@ -419,11 +419,20 @@ fn maskedEmbeddingArgmax(
...
 fn wantsMaskedEmbeddingArgmax(draft_cfg: gpt_mod.Config) bool {
+    if (getenvBool("ANTFLY_GEMMA4_MTP_DISABLE_MASKED_EMBEDDING")) return false;
     return draft_cfg.mtp_use_ordered_embeddings and
         draft_cfg.mtp_num_centroids != 0 and
         draft_cfg.mtp_centroid_intermediate_top_k != 0;
 }
 
+fn mtpPhaseNowNs() u64 {
+    var ts: std.posix.timespec = undefined;
+    switch (std.posix.errno(std.posix.system.clock_gettime(.MONOTONIC, &ts))) {
```
Comment (ajroetker) [inline zig/pkg/inference/src/architectures/gemma4_mtp.zig]: Should this be in or use lib/platform?

File: zig/pkg/inference/src/architectures/gemma4_mtp.zig [hunk 2]
```diff
@@ -928,6 +955,7 @@ pub fn draftTokenDevice(request: DraftDeviceRequest) !DraftDeviceResult {
     };
     profile.argmax_ns = profileElapsedNs(argmax_started_at);
 
+    if (phase_trace) { std.debug.print("mtp_phase: lm_argmax_us={d} source={s}\n", .{ (mtpPhaseNowNs() - phase_t0) / 1000, @tagName(logit_source) }); }
```
Comment (ajroetker) [inline zig/pkg/inference/src/architectures/gemma4_mtp.zig]: should this be on the profile instead of debug logging?

File: zig/pkg/inference/src/backends/decoder_gated_runtime.zig [hunk 3]
```diff
@@ -320,10 +320,27 @@ fn disableDirectGatedFamilyRequested() bool {
...
+fn disableGemma4DirectDecodeRequested() bool {
+    return getenvBool("TERMITE_METAL_DISABLE_GEMMA4_DIRECT_DECODE");
+}
+
 fn disableGemmaFusedQkvRequested() bool {
     return getenvBool("TERMITE_METAL_DISABLE_GEMMA_FUSED_QKV");
 }
 
+fn enableGemmaFusedQkvRequested() bool {
+    return getenvBool("TERMITE_METAL_ENABLE_GEMMA_FUSED_QKV");
```
Comment (ajroetker) [inline zig/pkg/inference/src/backends/decoder_gated_runtime.zig]: TODO: rename all these

File: zig/pkg/inference/build.zig [hunk 4]
```diff
@@ -353,6 +353,27 @@ pub fn build(b: *std.Build) void {
     );
     metal_prefill_bucket_bench_step.dependOn(&run_metal_prefill_bucket_bench.step);
 
+    const metal_q4_0_linear_bench_exe = b.addExecutable(.{
+        .name = "antfly-inference-metal-q4-0-linear-bench",
```
Comment (ajroetker) [inline zig/pkg/inference/build.zig]: does this need it's own bench target of should inference-metal-bench be it's own target with different options for filters

File: zig/pkg/inference/src/server/server.zig [hunk 5]
```diff
@@ -2553,6 +2557,12 @@ pub const Node = struct {
             .onnx => return ctx.status(500).json(.{ .@"error" = "BACKEND_ERROR", .message = "unexpected ONNX backend in native generation path" }),
             .wasm => return ctx.status(500).json(.{ .@"error" = "BACKEND_ERROR", .message = "unexpected WASM backend in server generation path" }),
         };
+        if (shouldDisableMetalAutoMtpDraft(config, model.session.backend(), platform.env.getenvBool("ANTFLY_GEMMA4_MTP_ENABLE_METAL_AUTO"))) {
```
Comment (ajroetker) [inline zig/pkg/inference/src/server/server.zig]: This one seems super specific and unneeded now
